### PR TITLE
Crypto refactoring and ECKey usage distinction

### PR DESCRIPTION
**Pull request by stevew817**
#### Problem
* Semantically different usage of the ECPKeypair's Serialize/Deserialize methods, where usage is mixed between converting to/from storage representation and to/from plaintext representation.
* No check against using an ECPKeypair for both ECDSA and ECDH, something which is warned against frequently

#### Change overview
* Rename ECPKey to ECPublicKey, ECPKeypair to ECKeypair and SupportedECPKeyTypes to SupportedECKeyTypes
* Refactor out intermediate abstract class P256KeypairBase since it had no use anywhere
* Refactor common class members to the base ECKeypair class instead of the top-level P256Keypair
* Introduced concept of `SupportedECKeyUsages` which restricts an ECKeypair to be used for either signing (ECDSA) or deriving (ECDH), and implemented checks in the presently shipping crypto backends. A default-initialised ECKeypair will be able to be used for signatures only.
* Refactored the semantic difference between plaintext keys and storage representation by introducing a new overloaded function `Initialize(P256SerializedKeypair)` which takes the role of 'constructing a P256Keypair object from a plaintext key'. This makes it easier to track and potentially refactor the remaining usages of `Serialize` and `Deserialize` later on, which now exclusively mean converting to/from storage representation. That will also facilitate implementing alternate crypto backends which come with their own storage implementation.

#### Testing
How was this tested? (at least one bullet point required)
* Refactor covered by existing testing
* Usage limitation tested by new testcases

### CODE REVIEW

**Comment by stevew817** [comment]
@tcarmelveilleux tagging you for your visibility since you seem to have been working on the crypto abstraction recently.

**Review by jmartinez-silabs — APPROVED** [review]

File: src/crypto/CHIPCryptoPAL.h
Comment (tcarmelveilleux) [inline src/crypto/CHIPCryptoPAL.h:356]: This usage change and the added template arguments **may cause major impact** to external users (who have had this stable API for a very long time)

Before this is approved, we need to determine if there is a way to resolve this without impact to users and this needs a  call-out to make sure we don't break assumptions.

**Review by tcarmelveilleux — CHANGES_REQUESTED** [review]
- This changes a core data structure in a way that may break very subtly the existing users who compile their code, but don't set the "Supported Key Usage" properly (and they didn't before).
- I see several files with the following comment:
```
    // Todo: support initialising a key from plaintext bytes when running
    // FabricTable / CASE unit tests with plaintext keys on a device with
    // this implementation
    return CHIP_ERROR_NOT_IMPLEMENTED;
```
  - We must make sure the above does not cause odd regressions.

**Comment by stevew817** [comment]
Abandoning this PR in favour of a less contentious solution after talking with @tcarmelveilleux offline (thanks for taking the time!).

**Pull request closed without merging**
